### PR TITLE
[DEVOPS-1083] Nix commenting and clean up

### PR DESCRIPTION
**Pull request by rvl**
## Description

The `default.nix` is very big and complicated. The build parameters need to be repeated everywhere (see #3757). There are a lack of comments explaining what the attributes and parameters are.

This branch does a general clean up and refactors the nix code to use the fixed point package set style which is used in nixpkgs.

## Linked issue

https://iohk.myjetbrains.com/youtrack/issue/DEVOPS-1083

## Type of change
- Refactor
- Documentation
- Build scripts and dev environment

## QA Steps
- `nix-build`
- Other QA steps TBD.

## TODO
- [ ] gather some initial feedback.
- [x] `release.nix`.
- [ ] more testing of build and dev environment.
- [ ] figure out how to pull the `cardanoPackages` attributes up to the top level without infinite recursion.
- [ ] get rid of `fetchNixpkgs.nix`

### CODE REVIEW

**Comment by angerman** [comment]
Ok, I've only looked over this superficially so far, I'll plan to look some more into this tomorrow.
- I really appreciate the removal of `localLib`.
- For the uninitiated, it might make sense to preamble each file with a high level overview of what the file is supposed to do.

File: lib.nix [hunk 1]
```diff
@@ -3,7 +3,11 @@ let
   fetchNixPkgs = let try = builtins.tryEval <cardano_pkgs>;
     in if try.success
     then builtins.trace "using host <cardano_pkgs>" try.value
-    else import ./fetch-nixpkgs.nix;
+    else import ./nix/fetch-nixpkgs.nix;
```
Comment (cleverca22) [inline lib.nix]: `fetch-nixpkgs.nix` is only a hack to allow `nix-build` in the buildkite hooks
the nix scripts should always be using `fetchNixpkgs.nix`

File: node-ipc/shell.nix [hunk 2]
```diff
@@ -1,3 +1,3 @@
-with import (import ../fetch-nixpkgs.nix) {};
+with import (import ../nix/fetch-nixpkgs.nix) {};
```
Comment (cleverca22) [inline node-ipc/shell.nix]: `fetchNixpkgs.nix` again

File: scripts/prepare-genesis/default.nix [hunk 3]
```diff
@@ -1,39 +1,39 @@
...
-  python = pkgs.python3.withPackages (ps: [ ps.pyyaml ]);
+  python = python3.withPackages (ps: [ ps.pyyaml ]);
 
-  yaml2json = pkgs.writeScriptBin "yaml2json" ''
+  yaml2json = writeScriptBin "yaml2json" ''
     #!${python}/bin/python
     import json
     import sys
     import yaml
     json.dump(yaml.load(open(sys.argv[1])), sys.stdout)
```
Comment (cleverca22) [inline scripts/prepare-genesis/default.nix]: `pkgs.haskell.lib.disableCabalFlag pkgs.haskellPackages.yaml "no-exe"` will give you a derivation with `yaml2json` and `json2yaml` binaries, which do the exact same job

File: lib.nix (hunk 1 above)
Comment (jmitchell) [reply lib.nix]: @cleverca22, is this distinction documented anywhere?

It's somewhat confusing there are two competing approaches that are pronounced the same way. Renaming one would help.

**Comment by rvl** [comment]
This was not ready to merge -- @angerman's review comments not addressed yet, there were unticked TODO items, the commit message said `wip`, and it fails to evaluate in Hydra.

**Pull request merged**